src/instructlab/sdg/llmblock.py

```diff
@@ -0,0 +1,107 @@
+# SPDX-License-Identifier: Apache-2.0
+# Standard
+import re
+
+# Third Party
+from datasets import Dataset
+
+# Local
+from .block import Block
+from .logger_config import setup_logger
+
+logger = setup_logger(__name__)
+
+
+class LLMBlock(Block):
+    def __init__(
+        self,
+        block_name,
+        config_path,
+        client,
+        model_id,
+        output_cols,
+        model_prompt="{prompt}",
+        **batch_kwargs,
+    ) -> None:
+        super().__init__(block_name)
+        self.block_config = self._load_config(config_path)
+        self.prompt_struct = (
+            """{system}\n{introduction}\n{principles}\n{examples}\n{generation}"""
+        )
+        self.prompt_template = self.prompt_struct.format(**self.block_config)
+        self.client = client
+        self.model = model_id
+        self.model_prompt = model_prompt
+        self.output_cols = output_cols
+        self.batch_params = batch_kwargs.get("batch_kwargs", {})
+        self.defaults = {
+            "model": self.model,
+            "temperature": 0,
+            "max_tokens": 12000,
+        }
+
+    def _parse(self, generated_string) -> dict:
+        matches = {}
+        for start_tag, end_tag, output_col in zip(
+            self.block_config["start_tags"],
+            self.block_config["end_tags"],
+            self.output_cols,
+        ):
+            if not start_tag and not end_tag:
+                matches[output_col] = (
+                    generated_string.strip() if generated_string else None
+                )
+            else:
+                pattern = re.escape(start_tag) + r"(.*?)" + re.escape(end_tag)
+                all_matches = re.findall(pattern, generated_string, re.DOTALL)
+                matches[output_col] = (
+                    [match.strip() for match in all_matches] if all_matches else None
+                )
+
+        return matches
+
+    def _generate(self, samples, **gen_kwargs) -> list:
+        prompts = [
```

Comment:
You can leave this as-is but in the future I recommend using a simpler for loop vs. list comprehension here so that it's simpler to read, e.g.:

```py
prompts = []
for sample in samples:
  prompt = self.prompt_template.format(**sample).strip()
  prompts.append(self.model_prompt.format(prompt))
```
Is easier to read than 

```py
        prompts = [
            self.model_prompt.format(
                prompt=self.prompt_template.format(**sample).strip()
            )
            for sample in samples
```

Reply:
I buy your point on readability, but I think list comprehensions are better in terms of performance